pandas/core/indexing.py

```diff
@@ -669,6 +670,56 @@ def _get_setitem_indexer(self, key):
 
         return self._convert_to_indexer(key, axis=0)
 
+    @final
+    def _maybe_mask_setitem_value(self, indexer, value):
+        """
+        If we have obj.iloc[mask] = arraylike and arraylike has the same
```

Comment:
I think you mean loc here?

Reply:
I think arraylike is not correct, we do this only when the rhs has an index. We align not apply the mask too

```
df = pd.DataFrame({"a": [1, 2, 3], "b": 1})

df.loc[[False, False, True], "b"] = np.ndarray([10, 11, 12])
```

Raises, which is what I would have expected